### PR TITLE
Track lastKnownRemoteTextEditingValue separately from received data

### LINKED ISSUE #48608: TextField calls InputFormatter twice for each keypress.
Extracted from b/146389916

The TextField calls the InputFormatter twice for each keypress. If the first pass changes the TextEditingValue state, then the second pass may produce different results as the first pass, resulting in unexpected behavior with no obvious cause.

This has been verified for when the keyboard type is set to `TextInputType.numberWithOptions()`

We should figure out why the second pass is called and see if we can just do one call of the formatter.

### PR DESCRIPTION
**Pull request by GaryQian**
# Description

Here, we track the last formatted value separately from the last passed-in update value to avoid double-formatting the update value.

Double formatting, depending on the implementation of the formatter, can produce unpredictable and incorrect behavior. 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/48608 and
https://b.corp.google.com/issues/144323703

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
It looks like this is causing some existing tests to fail.

**Comment by chunhtai** [comment]
Do you figure out why it only happens on some devices?

**Review by justinmc — APPROVED** [review]
LGTM, sorry I missed this at first!

**Comment by goderbauer** [comment]
Looks like Cirrus is unhappy with this change.

**Comment by GaryQian** [comment]
Working to resolve an edge case where non-bugged keyboards can skip a format cycle on repeat keypresses.

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 1]
```diff
@@ -4178,6 +4178,113 @@ void main() {
...
+    state.updateEditingValue(const TextEditingValue(text: 'aa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aaa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aaaa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aaaaaaa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aaaaaaaaa'));
+    state.updateEditingValue(const TextEditingValue(text: 'aaaaaaaaa')); // Skipped
+
+    final List<String> referenceLog = <String>[];
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/editable_text_test.dart]: Why not just a const list of literals?  Seems a lot more readable.

```suggestion
const List<String> referenceLog = <String>[
  '[1]: , a',
  '[1]: normal aa',
  '[2]: aa, aaa',
  '[2]: normal aaaa',
  '[3]: aaaa, aa',
  '[3]: deleting a',
  '[4]: a, aaa',
  '[4]: normal aaaaaaaa',
  '[5]: aaaaaaaa, aaaa',
  '[5]: deleting aaa',
  '[6]: aaa, aa',
  '[6]: deleting aaaa',
  '[7]: aaaa, aaaaaaa',
  '[7]: normal aaaaaaaaaaaaaa',
  '[8]: aaaaaaaaaaaaaa, aa',
  '[8]: deleting aaaaaa',
  '[9]: aaaaaa, aaaaaaaaa',
  '[9]: normal aaaaaaaaaaaaaaaaaa',
];
```

Comment (GaryQian) [reply packages/flutter/test/widgets/editable_text_test.dart]: Oh yeah that makes sense.

**Pull request merged**